### PR TITLE
[GOES TO FEATURE BRANCH] Main article for migrating to service catalog from ref arch v1

### PR DESCRIPTION
**Pull request by yorinasub17**
__This merges into feature branch `yori-refarch2-mig`__

This is the main migration article for updating from the old v1 reference architecture to v2 using the service catalog. Note that the following pieces are still missing, and will be added in incrementally:

- The migration scripts
- The starter pack zip file link
- Dedicated service module migration guides

The service module migration guides will be incrementally ported over as they are done in separate PRs. Hence, the approach to use a feature branch.

### CODE REVIEW

**Review by bwhaley — APPROVED** [review]

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (brikis98) [inline _posts/2021-05-04-how-to-update-your-ref-arch.adoc:195]: "In Ref Arch 1.0, we had ..."

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (yorinasub17) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:2]: The reason it's not focusing on Ref Arch 2.0 is because we kicked out Gruntwork Pipelines and LZ migration as out of scope. Gruntwork Pipelines ends up being a deploy from scratch situation (which is covered by https://gruntwork.io/guides/automations/how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code/), and the amount of effort involved in resolving all the imports to make LZ work felt like it was too large in scope that it will significantly extend the timeline.

Note that this is largely driven by the issue that opening up service catalog to all customers is blocked on this guide, and the longer we go with the service catalog closed, the longer we have to maintain 3 copies of landing zone (`terraform-aws-security`, `terraform-aws-service-catalog`, and `terraform-aws-cis-service-catalog`).

Given that, this guide is really only focused on using Service Catalog + `dependency` blocks.

Comment (yorinasub17) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:2]: Ok despite what I said, here is what I propose for handling this:

- Make the update to scope this as a v1 => v2 upgrade and describe the differences as you mentioned.
- Add two help boxes: one indicating that the landing zone migration guide will be published later, and one indicating that the steps for adding Gruntwork Pipelines is in the Gruntwork Pipelines deployment guide.

What do you think?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (yorinasub17) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc]: > NIT: for SEO, it's better to put the link on the relevant words, such as Service Catalog.

Done! 38f0817f

> Also, we should have an announcement blog post for Ref Arch 2.0, and that's what we should link to and discuss here in the background section.

FYI we have a ticket to add this (https://gruntwork.atlassian.net/browse/IAC-1831), so we are on the same page!

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (yorinasub17) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc]: Done in 38f0817

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bwhaley) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:2]: > one indicating that the landing zone migration guide will be published later,

Can we just point to the existing Landing Zone guide? I'd hate to create a future work stream if it isn't necessary. I guess you're thinking that a migration guide would cover how to get from the legacy set up where some of the account-baseline functionality (like cloudtrail and cross-account-iam-roles) are in their own, dedicated modules. Or do I misunderstand what the landing zone migration guide would cover?

Comment (brikis98) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:2]: Agreed on the proposed changes, including Ben's note!

Comment (yorinasub17) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:2]: > I guess you're thinking that a migration guide would cover how to get from the legacy set up where some of the account-baseline functionality (like cloudtrail and cross-account-iam-roles) are in their own, dedicated modules. Or do I misunderstand what the landing zone migration guide would cover?

This is what I am thinking. A way to import your existing fragmented landing zone modules (`iam-groups`, `cross-account-iam-roles`, etc) into the `account-baseline-app` module. The landing zone deployment guide assumes a blank account, so it is useful for new accounts, but it won't be useful for customers who are migrating their existing account to landing zone (as the roles and stuff will conflict).

Hence, I don't think we should even point customers to the landing zone deployment guide because that would lead customers down the wrong path.

**Comment by rhoboat** [comment]
I noticed a couple of tiny things, and I've also got ideas for improvements. Since you're out for a week, do you mind if I carry this PR forward? @yorinasub17 @bwhaley @brikis98 ?

**Comment by brikis98** [comment]
Fine by me! I believe the CIS Ref Arch is good to go, so I think wrapping up Ref Arch 2.0 and migration guides is now the highest priority!

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:2]: My thought was that we'd announce the new ref arch in the blog, and link from that announcement to this migration guide. We can describe all the differences and promote it there, I feel.

**Comment by bendavies** [comment]
Super! As someone on ref arch v1, I've been looking forward this for a while.
I'll give it a good read tomorrow to see if it all makes sense to me.

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bendavies) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:2]: excuse my ignorance, but what is "Ref Arch 2.0" is there some additional step to take from "Ref Arch 1.0" -> "Service Catalog" -> "Ref Arch 2.0" ?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bendavies) [inline _posts/2021-05-04-how-to-update-your-ref-arch.adoc:135]: why doesn't this use `read_terragrunt_config` as well?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bendavies) [inline _posts/2021-05-04-how-to-update-your-ref-arch.adoc:256]: this is a pretty easy step to follow, so just make it non optional?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bendavies) [inline _posts/2021-05-04-how-to-update-your-ref-arch.adoc:135]: this is a pretty easy step to follow, so just make it non optional?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bendavies) [inline _posts/2021-05-04-how-to-update-your-ref-arch.adoc:269]:
```suggestion
#!/usr/bin/env bash
```

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bendavies) [inline _posts/2021-05-04-how-to-update-your-ref-arch.adoc:424]: this is not the latest version. there are 2 releases after this. what if we are on a later version?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bendavies) [inline _posts/2021-05-04-how-to-update-your-ref-arch.adoc:431]: a full list of `corresponding Terraform module` mappings might be beneficial here?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bendavies) [inline _posts/2021-05-04-how-to-update-your-ref-arch.adoc:461]: should this be?
```suggestion
.. *Make a backup of the state file* using the following command: `terragrunt state pull --terragrunt-source ... > backup.tfstate`. You can use this to rollback the state to before you attempted the migration with the
   following command: `terragrunt state push --terragrunt-source ..."$(pwd)/backup.tfstate"`.
```

also, what should `--terragrunt-source` be?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:48]: For the purpose of this migration guide, I think we want to scope core-concepts to the main difference here, which is the Service Catalog. I think we should write up a separate announcement article that describes that stuff in more detail. What are your thoughts?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:424]: Newer is fine. Updating the text to reflect that. Done in a7b778a6..549c027e.

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc]: This is indeed misplaced. Users don't need this. Thanks for catching! 549c027e..e76eb55f

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc]: We've got a different approach to take here. Each dedicated guide will have that info.

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:255]: I don't feel too strongly about it, but weighing it, I think we should keep it optional, in case people don't want to do this. Yes, it's easy, but I don't want to give the impression that it's required. ⚖️ I could go either way on it, really.

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:268]: I'd like to incorporate this suggestion, but we need to also change the corresponding line of code in the provided zip file.

Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:268]: Done in 427c99cd..8d2468ff, also will do in the zip file.

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:134]: Same feelings as above. Even if it's easy, it's not required, so I'm leaning toward keeping it optional.

**Comment by rhoboat** [comment]
Latest changes awaiting another review: https://github.com/gruntwork-io/gruntwork-io.github.io/pull/416/files/7c3acab38603b3865fadd436646eff3385404936..3fe804c4ed5fe51a9c8ff7f27ac6745e491a062a

* Added tip boxes. Suggestions for better wording appreciated.
* Tried to add a link to the folder structure. I think this shoudl actually be a release on the service-catalog. Will fix this.
* Misc changes
* Point to `v0.35.5` of the service-catalog now.

**Comment by bendavies** [comment]
what is the `_regional` folder in the `refarch-folder-structure` zip? this is new to me.

**Comment by bendavies** [comment]
is it worth including a link to https://github.com/gruntwork-io/infrastructure-live-multi-account-acme/pull/47 somewhere in here, as it provides a good example of how to complete the migration

**Comment by bwhaley** [comment]
> what is the _regional folder in the refarch-folder-structure zip? this is new to me.
@bendavies this is just a more accurate name for what used to be `_global`. It indicates services that are _regional_ in nature, but not _global_ to an account. For example, IAM is a _global_ service, but SNS, Secrets Manager, and KMS are _regional_ services.

**Comment by bwhaley** [comment]
> is it worth including a link to gruntwork-io/infrastructure-live-multi-account-acme#47 somewhere in here, as it provides a good example of how to complete the migration

That's a good point! I think we should add this link if we haven't already.

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bwhaley) [inline _posts/2021-05-04-how-to-update-your-ref-arch.adoc:38]: I'm personally not keen on writing a Landing Zone migration guide. Customers that want to adopt the Landing Zone can follow [this guide](https://gruntwork.io/guides/foundations/how-to-configure-production-grade-aws-account-structure/). But I think others might feel differently from me on this matter.

**Review by bwhaley — APPROVED** [review]
Left a couple of comments, but it LGTM!

**Comment by bendavies** [comment]
> > what is the _regional folder in the refarch-folder-structure zip? this is new to me.
> 
> @bendavies this is just a more accurate name for what used to be `_global`. It indicates services that are _regional_ in nature, but not _global_ to an account. For example, IAM is a _global_ service, but SNS, Secrets Manager, and KMS are _regional_ services.

Thanks. The rename should be detailed in this guide then I assume?

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (bwhaley) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:2]: Ref Arch 2.0 is the name we've been using for the refactored reference architecture that uses the service catalog and incorporates the CI/CD pipeline and the Landing Zone. It's automatically generated from the architecture catalog, which we haven't yet released to customers. It's depicted in our marketing copy here: https://gruntwork.io/reference-architecture/

**Comment by bwhaley** [comment]
> Thanks. The rename should be detailed in this guide then I assume?

Probably worth mentioning! /cc @rhoboat

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:4]: Good suggestion, however, not doing. Editing and creating a new icon is too much of a distraction at this time.

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:48]: We're moving forward with maintaining the scope of this article to the migration only. Thanks!

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc]: Yeah.. I think actually I've never really seen "end-user" hyphenated.

**Comment by rhoboat** [comment]
Thanks for the review! Merging now _to the feature branch_. Taking care of these items in the next iteration:

* Each guide needs to list dependencies in a section at the top. Would appreciate help getting this done.
* Each guide could link to the corresponding patch/diff from this PR https://github.com/gruntwork-io/infrastructure-live-multi-account-acme/pull/47
* EKS guides need to be patched/diffed as per that PR also.
* Guides need to be converted md -> adoc (I use pandoc).
* **Assets**. 
    * The dedicated guides can be in a subfolder where the main guide is. 
    * The folder-structure.zip is going to be a release on terraform-aws-service-catalog
    * The migration scripts should be copied to the website in a subfolder, so they can be linked from the guide

File: _posts/2021-05-04-how-to-update-your-ref-arch.adoc
Comment (rhoboat) [reply _posts/2021-05-04-how-to-update-your-ref-arch.adoc:2]: @bendavies Good question, and I didn't realize this could be a point of confusion. I'll (or whoever writes it will) point out in our announcement that the Service Catalog Ref Arch _is_ Ref Arch 2.0.

**Pull request merged**
